presto-main/src/main/java/com/facebook/presto/operator/unnest/UnnestOperatorBlockUtil.java

```diff
@@ -13,32 +13,304 @@
  */
 package com.facebook.presto.operator.unnest;
 
-import static java.lang.String.format;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.ArrayBlock;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.ByteArrayBlock;
+import com.facebook.presto.spi.block.ColumnarArray;
+import com.facebook.presto.spi.block.ColumnarMap;
+import com.facebook.presto.spi.block.ColumnarRow;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.Int128ArrayBlock;
+import com.facebook.presto.spi.block.IntArrayBlock;
+import com.facebook.presto.spi.block.LongArrayBlock;
+import com.facebook.presto.spi.block.MapBlock;
+import com.facebook.presto.spi.block.RowBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import com.facebook.presto.spi.block.ShortArrayBlock;
+import com.facebook.presto.spi.block.VariableWidthBlock;
+import com.facebook.presto.spi.function.OperatorType;
+import com.facebook.presto.spi.type.MapType;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
+import static com.facebook.presto.spi.block.ArrayBlock.fromElementBlock;
+import static com.facebook.presto.spi.block.MapBlock.fromKeyValueBlock;
+import static com.facebook.presto.spi.block.MethodHandleUtil.compose;
+import static com.facebook.presto.spi.block.MethodHandleUtil.nativeValueGetter;
+import static com.facebook.presto.spi.block.RowBlock.fromFieldBlocks;
+import static com.facebook.presto.testing.TestingEnvironment.TYPE_MANAGER;
 
 final class UnnestOperatorBlockUtil
 {
     private UnnestOperatorBlockUtil() {}
 
-    private static final int DEFAULT_CAPACITY = 64;
-    // See java.util.ArrayList for an explanation
-    static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
+    static Block insertNull(Block block, Type type)
+    {
+        if (block instanceof DictionaryBlock) {
```

Comment:
Actually I'm not sure why this class is needed.

It doesn't provide any bulk-copy functionality. Why not simply create a new block builder by calling the `Type#createBlockBuilder` and copy the block into a new block builder?

Also we can have a `Type#createBlockBuilder(Block ...)` that creates a block builder based on the copy of block (and does that copy more efficient if it is proven that reconstructing the block using a BlockBuilder negatively impacts the performancce)

Reply:
@arhimondr  Copying the block using these tight loops is 5-6x faster than calling `type.appendTo(block, j, blockBuilder)` or `blockBuilder.writeXXX(block.getXXX(i))` therefore I call it bulk copy. 

Let me use an example to see if I understand you correctly: You suggested adding `Type#createBlockBuilder(..., Block block)`, and add a new constructor to each BlockBuilder. E.g.
AbstractLongType:
```
@Override
    public final BlockBuilder createBlockBuilder(Block block, BlockBuilderStatus blockBuilderStatus)
    {
        int maxBlockSizeInBytes;
        if (blockBuilderStatus == null) {
            maxBlockSizeInBytes = PageBuilderStatus.DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
        }
        else {
            maxBlockSizeInBytes = blockBuilderStatus.getMaxPageSizeInBytes();
        }
        return new LongArrayBlockBuilder(
                blockBuilderStatus,
                block);
    }
```
Then add a new constructor of `LongArrayBlockBuilder`
```
    public LongArrayBlockBuilder(@Nullable BlockBuilderStatus blockBuilderStatus, Block block)
    {
        this.blockBuilderStatus = blockBuilderStatus;

        int positionCount = block.getPositionCount();
        values = new long[positionCount + 1];
        for (int i = 0; i < positionCount; i++) {
            values[i] = block.getLong(i);
        }

        valueIsNull = copyIsNulls(block);

        updateDataSize();
    }
```
Is this correct?

This should work well, but it still requires copying. I'd prefer adding a appendNull() to Block, so that in many places we could avoid copying at all. It was my first choice but I was afraid it would be more controversial than doing memory copy. But since @mbasmanova also suggested it I'd do it that way. Are you ok with it?